pkg/deploy/strategy/support/lifecycle.go

```diff
@@ -239,100 +239,107 @@ func NewFirstContainerReady(kclient kclient.Interface, timeout time.Duration, in
+	glog.V(0).Infof("Waiting %.f seconds for pods owned by deployment %q to become ready (checking every %.f seconds; %d pods previously accepted)", c.timeout.Seconds(), deployutil.LabelForDeployment(deployment), c.interval.Seconds(), c.acceptedPods.Len())
+	err := wait.Poll(c.interval, c.timeout, func() (done bool, err error) {
+		// Update the observed ready status of all pod containers.
+		for _, obj := range podStore.List() {
+			pod := obj.(*kapi.Pod)
+			// Skip previously accepted pods; we only want to verify newly observed
+			// pods on each overall acceptance pass.
+			if c.acceptedPods.Has(pod.Name) {
 				continue
 			}
-			// The status of the container could be transient; we only care if it
-			// was ever ready. If it was ready and then became not ready, we
-			// consider it ready.
-			if status.Ready {
-				observedContainers[status.Name] = true
+			// If this is the first time we've seen the pod, consider all its
+			// containers to be unready.
+			if _, exists := observedPodContainers[pod.Name]; !exists {
+				observedPodContainers[pod.Name] = map[string]bool{}
+				for _, container := range pod.Spec.Containers {
+					observedPodContainers[pod.Name][container.Name] = false
+				}
+			}
+			// The status of a container could be transient; we only care if it was
+			// ever ready. If it was observed to be ready and then became unready,
+			// we consider it ready.
+			for _, status := range pod.Status.ContainerStatuses {
+				if status.Ready {
+					observedPodContainers[pod.Name][status.Name] = true
+				}
 			}
 		}
-		// Check whether all containers have been observed as ready.
-		allReady := true
-		for _, ready := range observedContainers {
-			if !ready {
-				allReady = false
-				break
+		// Check whether all containers have been observed as ready and collect
+		// some basic stats to report.
+		unreadyPods := 0
+		unreadyContainers := 0
+		for _, containers := range observedPodContainers {
+			allContainersReady := true
+			for _, ready := range containers {
+				if !ready {
+					unreadyContainers++
+					allContainersReady = false
+				}
+			}
+			if !allContainersReady {
+				unreadyPods++
 			}
 		}
 		// If all containers have been ready once, return success.
-		if allReady {
-			glog.V(0).Infof("All containers ready for %s/%s", pod.Namespace, pod.Name)
+		if unreadyPods == 0 {
+			glog.V(0).Infof("Finished; %d new pods ready for %s", len(observedPodContainers), deployutil.LabelForDeployment(deployment))
 			return true, nil
 		}
 		// Otherwise, try again later.
-		glog.V(4).Infof("Still waiting for pod %s/%s container readiness; observed statuses: #%v", pod.Namespace, pod.Name, observedContainers)
+		glog.V(4).Infof("Still waiting for %d of %s pods to become ready for deployment %s", len(observedPodContainers), unreadyPods, unreadyContainers, deployutil.LabelForDeployment(deployment))
```

Comment:
We should spawn a follow up on the usability  of output.  Let's get this
closedz

On Jul 21, 2015, at 9:11 AM, Dan Mace <notifications@github.com> wrote:

In pkg/deploy/strategy/support/lifecycle.go
<https://github.com/openshift/origin/pull/3701#discussion_r35099349>:

>  			return true, nil
>  		}
>  		// Otherwise, try again later.
> -		glog.V(4).Infof("Still waiting for pod %s/%s container readiness; observed statuses: #%v", pod.Namespace, pod.Name, observedContainers)
> +		glog.V(4).Infof("Still waiting for %d of %s pods to become ready for deployment %s", len(observedPodContainers), unreadyPods, unreadyContainers, deployutil.LabelForDeployment(deployment))

I really wanted to emit progress at V(0) but I'm not sure of a
great/all-purpose way to rate limit the output when the interval is short.
Thoughts?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701/files#r35099349>.